File: backend/routes/executions.py
from fastapi import APIRouter, HTTPException
from repositories.executions import execution_repo
from utils.response import ApiResponse

router = APIRouter(prefix="/executions", tags=["Executions"])


@router.get("", response_model=ApiResponse)
async def get_executions(task_id: int = None, agent_id: int = None):
    """Get all executions, optionally filtered by task or agent"""
    if task_id:
        executions = execution_repo.get_by_task(task_id)
    elif agent_id:
        executions = execution_repo.get_by_agent(agent_id)
    else:
        executions = execution_repo.get_all()
    return ApiResponse.ok(data=executions)


@router.get("/{execution_id}", response_model=ApiResponse)
async def get_execution(execution_id: int):
    """Get an execution by ID"""
    execution = execution_repo.get_by_id(execution_id)
    if not execution:
        raise HTTPException(status_code=404, detail="Execution not found")
    return ApiResponse.ok(data=execution)


@router.post("", response_model=ApiResponse)
async def create_execution(execution_in: dict):
    """Create a new execution"""
    execution = execution_repo.create(execution_in)
    return ApiResponse.ok(
        data=execution, message="Execution started"
    )


@router.put("/{execution_id}", response_model=ApiResponse)
async def update_execution(
    execution_id: int,
    execution_in: dict,
):
    """Update an execution"""
    execution = execution_repo.get_by_id(execution_id)
    if not execution:
        raise HTTPException(status_code=404, detail="Execution not found")
    updated = execution_repo.update(execution_id, execution_in)
    return ApiResponse.ok(
        data=updated, message="Execution updated"
    )


@router.post("/{execution_id}/stop", response_model=ApiResponse)
async def stop_execution(execution_id: int):
    """Stop an execution"""
    execution = execution_repo.get_by_id(execution_id)
    if not execution:
        raise HTTPException(status_code=404, detail="Execution not found")
    updated = execution_repo.update(execution_id, {"status": "CANCELLED"})
    return ApiResponse.ok(
        data=updated, message="Execution stopped"
    )


@router.delete("/{execution_id}", response_model=ApiResponse)
async def delete_execution(execution_id: int):
    """Delete an execution"""
    execution = execution_repo.get_by_id(execution_id)
    if not execution:
        raise HTTPException(status_code=404, detail="Execution not found")
    execution_repo.delete(execution_id)
    return ApiResponse.ok(message="Execution deleted")

File: backend/src/routes/executions.ts
import type { FastifyPluginAsync } from 'fastify';

import { ExecutionService } from '../services/executionService.js';
import type { CreateExecutionInput, UpdateExecutionInput } from '../types/dto/executions.js';
import type { IdParams, SessionIdParams, TaskIdParams } from '../types/http/params.js';
import { successResponse, errorResponse } from '../utils/response.js';
import { getErrorMessage, getStatusCode, parseNumber, logError } from '../utils/http.js';

const executionService = new ExecutionService();

export const executionRoutes: FastifyPluginAsync = async (fastify) => {
  fastify.get('/', async (request) => {
    try {
      return successResponse(await executionService.getAll());
    } catch (error) {
      logError(error, request);
      return errorResponse('Failed to get executions');
    }
  });

  fastify.get<{ Params: IdParams }>('/:id', async (request, reply) => {
    try {
      const execution = await executionService.getById(parseNumber(request.params.id));
      if (!execution) {
        reply.code(404);
        return errorResponse('Execution not found');
      }
      return successResponse(execution);
    } catch (error) {
      logError(error, request);
      reply.code(getStatusCode(error));
      return errorResponse(getErrorMessage(error, 'Failed to get execution'));
    }
  });

  fastify.get<{ Params: SessionIdParams }>('/session/:sessionId', async (request) => {
    try {
      return successResponse(await executionService.getBySession(parseNumber(request.params.sessionId)));
    } catch (error) {
      logError(error, request);
      return errorResponse('Failed to get executions');
    }
  });

  fastify.get<{ Params: TaskIdParams }>('/task/:taskId', async (request) => {
    try {
      return successResponse(await executionService.getByTask(parseNumber(request.params.taskId)));
    } catch (error) {
      logError(error, request);
      return errorResponse('Failed to get executions');
    }
  });

  fastify.post<{ Body: CreateExecutionInput }>('/', async (request, reply) => {
    try {
      const execution = await executionService.create(request.body || { session_id: 0 });
      return successResponse(execution, 'Execution created');
    } catch (error) {
      logError(error, request);
      const statusCode = getStatusCode(error);
      if (statusCode === 404) {
        reply.code(404);
        return errorResponse(getErrorMessage(error, 'Session not found'));
      }
      reply.code(getStatusCode(error));
      return errorResponse(getErrorMessage(error, 'Failed to create execution'));
    }
  });

  fastify.put<{ Params: IdParams; Body: UpdateExecutionInput }>('/:id', async (request, reply) => {
    try {
      const updated = await executionService.update(parseNumber(request.params.id), request.body);
      if (!updated) {
        reply.code(404);
        return errorResponse('Execution not found');
      }
      return successResponse(updated, 'Execution updated');
    } catch (error) {
      logError(error, request);
      reply.code(getStatusCode(error));
      return errorResponse(getErrorMessage(error, 'Failed to update execution'));
    }
  });

  fastify.delete<{ Params: IdParams }>('/:id', async (request, reply) => {
    try {
      const deleted = await executionService.delete(parseNumber(request.params.id));
      if (!deleted) {
        reply.code(404);
        return errorResponse('Execution not found');
      }
      return successResponse(null, 'Execution deleted');
    } catch (error) {
      logError(error, request);
      reply.code(getStatusCode(error));
      return errorResponse(getErrorMessage(error, 'Failed to delete execution'));
    }
  });
};
